Fix part of #7712: Added an audit job to validate continue button's length.

iamprayush opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes part of #7712.
2. This PR does the following: We need to enforce a length limit on the continue button's label text. This PR adds a one-off job to check if all explorations are under the limit (50).

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

iamprayush committed:
Added the audit job.

DubeySandeep commented on core/domain/interaction_jobs_one_off.py:
```diff
@@ -212,3 +213,36 @@ def map(item):
+
+    @staticmethod
+    def map(item):
+        if not item.deleted:
```
```
If item.deleted:
    return
```

iamprayush replied:
Done.

DubeySandeep commented on core/domain/interaction_jobs_one_off_test.py:
```diff
@@ -874,3 +874,124 @@ def test_no_action_is_performed_for_deleted_exploration(self):
+        # Start ItemSelectionInteractionOneOff job on sample exploration.
+        job_id = (
+            interaction_jobs_one_off
+            .InteractionCustomizationArgsValidationOneOffJob.create_new())
```
WHy are we running `InteractionCustomizationArgsValidationOneOffJob` job? [Ditto elsewhere in this file!]

iamprayush replied:
MB, I've fixed it now.

DubeySandeep commented on core/domain/interaction_jobs_one_off_test.py:
```diff
@@ -874,3 +874,124 @@ def test_no_action_is_performed_for_deleted_exploration(self):
+        run_job_for_deleted_exp(
+            self,
+            interaction_jobs_one_off
+            .InteractionCustomizationArgsValidationOneOffJob)
```
Add a test with state's continue interaction having more than 50 chars.

iamprayush replied:
I have added it [here](https://github.com/oppia/oppia/blob/17ef154b53eec08c580a3c879887bbbb7513054d/core/domain/interaction_jobs_one_off_test.py#L945). Did I miss something?
Also, I've added tests for less than 50 and greater than 50 and checked than only the latter is in the output. That should be enough, right?

DubeySandeep replied:
I was confused as the job output was `self.assertEqual(actual_output, [])` everywhere. Thanks for the info!

DubeySandeep reviewed (changes requested):
@iamprayush, Thanks for the PR, the job looks good to me in general. I think you need to write better test, I've left a few comments PTAL!

seanlip reviewed (commented):
LGTM for jobs_registry.py (codeowners).

@iamprayush Please file a job run request in the usual form so that this can be tested; thanks!

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into audit-label-overflow

iamprayush committed:
Fixed tests.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into audit-label-overflow

iamprayush commented:
> @iamprayush Please file a job run request in the usual form so that this can be tested; thanks!

I've filled the form, PTAL!

seanlip commented:
Hi @iamprayush -- this is running currently, but please go ahead and update the PR (or make a new one) that enforces this validation check. In your release instructions, explain that the audit job should be run and any issues fixed **prior to deployment**. That should lead to a clean enforcement of this check. Thanks!

seanlip commented:
Also, the job fails. Log:

```
Traceback (most recent call last): (/third_party/gae-mapreduce-1.9.22.0/mapreduce/handlers.py:517)
  File "/third_party/gae-mapreduce-1.9.22.0/mapreduce/handlers.py", line 503, in handle
    tstate.input_reader, shard_state, tstate, ctx)
  File "/third_party/gae-mapreduce-1.9.22.0/mapreduce/handlers.py", line 593, in _process_inputs
    entity, input_reader, ctx, tstate):
  File "/third_party/gae-mapreduce-1.9.22.0/mapreduce/handlers.py", line 635, in _process_datum
    for output in result:
  File "/core/domain/interaction_jobs_one_off.py", line 245, in map
    python_utils.UNICODE(button_text_length)))
UnicodeDecodeError: 'ascii' codec can't decode byte 0xd8 in position 0: ordinal not in range(128)
```

~~A general tip: please use .encode('utf-8') following the other jobs. In general, look at other surrounding code, don't invent new things if what you're writing is similar to existing stuff.~~ [Scratch that -- I misread the error message. Probably more investigation is needed to figure out what's going wrong, perhaps try running it on your own dev server with unicode chars / emoji.]

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into audit-label-overflow

iamprayush committed:
Fixed unicode decode error.

iamprayush committed:
Removed unused import.

iamprayush commented:
Sorry about this, @seanlip. I have now fixed this. I realized that we don't need to convert the length (which is an integer) into a Unicode string (which is where the error was occurring) before interpolating it into the output string. A similar use-case is present [here](https://github.com/oppia/oppia/blob/develop/core/domain/activity_jobs_one_off.py#L311-L318). So, I believe we can be sure that this error won't recur now.

> Hi @iamprayush -- this is running currently, but please go ahead and update the PR (or make a new one) that enforces this validation check. In your release instructions, explain that the audit job should be run and any issues fixed prior to deployment. That should lead to a clean enforcement of this check. Thanks!

I've created a new PR(https://github.com/oppia/oppia/pull/10539) for that and I've updated the instructions doc.

seanlip commented:
Still throws an error:

```
Traceback (most recent call last): (/third_party/gae-mapreduce-1.9.22.0/mapreduce/handlers.py:517)
  File "/third_party/gae-mapreduce-1.9.22.0/mapreduce/handlers.py", line 503, in handle
    tstate.input_reader, shard_state, tstate, ctx)
  File "/third_party/gae-mapreduce-1.9.22.0/mapreduce/handlers.py", line 593, in _process_inputs
    entity, input_reader, ctx, tstate):
  File "/third_party/gae-mapreduce-1.9.22.0/mapreduce/handlers.py", line 635, in _process_datum
    for output in result:
  File "/core/domain/interaction_jobs_one_off.py", line 243, in map
    state_name.encode('utf-8'), button_text_length))
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 21: ordinal not in range(128)
```

Did you actually test manually with unicode? I didn't think it was due to the number interpolation. The stack trace just shows the last line of the call that failed, but you'll need to examine the whole call. (That was the error I made the first time.)

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into audit-label-overflow

iamprayush committed:
Fixed unicode decode error.

iamprayush committed:
Removed unused import

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into audit-label-overflow

iamprayush commented:
I see. Okay so I did a little more experimenting and I found out that this error was occurring because of the presence unicode characters in the state name, which is weird because the state name is interpolated like this (`state_name.encode('utf-8')`) everywhere else in the codebase.
Anyways, I have now fixed this and I've added a unicode character in the tests as well to confirm this fix.

DubeySandeep commented on core/domain/interaction_jobs_one_off.py:
```diff
@@ -212,3 +212,37 @@ def map(item):
+        exploration = exp_fetchers.get_exploration_from_model(item)
+        for state_name, state in exploration.states.items():
+            interaction = state.interaction
+            if interaction.id == 'Continue':
```
optional:
```suggestion
            if interaction.id != 'Continue':
                continue
```
This will help us avoid branching and have readable and fast code*

DubeySandeep reviewed (approved):
LGTM for the codeowner file*

seanlip commented:
Thanks @iamprayush, it ran! I think the lesson here is always to do a full run of the job locally with unicode everywhere before you can be certain it works (though @brianrodri has made a nice PR in #10561 that might help simplify this). I've sent you the output and will await further instructions (specifically, how to fix things up).
